java/src/main/native/ai_onnxruntime_OrtSession.c

```diff
@@ -13,227 +13,286 @@
  * Signature: (JJJ)[Ljava/lang/String;
  */
-JNIEXPORT jobjectArray JNICALL Java_ai_onnxruntime_OrtSession_getInputNames
-  (JNIEnv * jniEnv, jobject jobj, jlong apiHandle, jlong sessionHandle, jlong allocatorHandle) {
-    (void) jobj; // Required JNI parameter not needed by functions which don't need to access their host object.
-    const OrtApi* api = (const OrtApi*) apiHandle;
-    OrtAllocator* allocator = (OrtAllocator*) allocatorHandle;
-
-    // Setup
-    char *stringClassName = "java/lang/String";
-    jclass stringClazz = (*jniEnv)->FindClass(jniEnv, stringClassName);
-
-    // Get the number of inputs
-    size_t numInputs = Java_ai_onnxruntime_OrtSession_getNumInputs(jniEnv, jobj, apiHandle, sessionHandle);
-
-    // Allocate the return array
-    jobjectArray array = (*jniEnv)->NewObjectArray(jniEnv,safecast_size_t_to_jsize(numInputs),stringClazz,NULL);
-    for (uint32_t i = 0; i < numInputs; i++) {
-        // Read out the input name and convert it to a java.lang.String
-        char* inputName;
-        checkOrtStatus(jniEnv,api,api->SessionGetInputName((OrtSession*)sessionHandle, i, allocator, &inputName));
-        jstring name = (*jniEnv)->NewStringUTF(jniEnv,inputName);
-        (*jniEnv)->SetObjectArrayElement(jniEnv, array, i, name);
-        checkOrtStatus(jniEnv,api,api->AllocatorFree(allocator,inputName));
+JNIEXPORT jobjectArray JNICALL Java_ai_onnxruntime_OrtSession_getInputNames(JNIEnv* jniEnv, jobject jobj, jlong apiHandle, jlong sessionHandle, jlong allocatorHandle) {
+  (void)jobj;  // Required JNI parameter not needed by functions which don't need to access their host object.
+  const OrtApi* api = (const OrtApi*)apiHandle;
+  OrtAllocator* allocator = (OrtAllocator*)allocatorHandle;
+  OrtSession* session = (OrtSession*)sessionHandle;
+
+  // Setup
+  char* stringClassName = "java/lang/String";
+  jclass stringClazz = (*jniEnv)->FindClass(jniEnv, stringClassName);
+
+  // Get the number of inputs
+  size_t numInputs = 0;
+  OrtErrorCode code = checkOrtStatus(jniEnv, api, api->SessionGetInputCount(session, &numInputs));
+  if (code != ORT_OK) {
+    return NULL;
+  }
+
+  // Allocate the return array
+  jobjectArray array = (*jniEnv)->NewObjectArray(jniEnv, safecast_size_t_to_jsize(numInputs), stringClazz, NULL);
+  for (uint32_t i = 0; i < numInputs; i++) {
+    // Read out the input name and convert it to a java.lang.String
+    char* inputName = NULL;
+    code = checkOrtStatus(jniEnv, api, api->SessionGetInputName(session, i, allocator, &inputName));
+    if (code != ORT_OK) {
+      // break out on error, return array and let Java throw the exception.
+      break;
+    }
+    jstring name = (*jniEnv)->NewStringUTF(jniEnv, inputName);
+    (*jniEnv)->SetObjectArrayElement(jniEnv, array, i, name);
+    code = checkOrtStatus(jniEnv, api, api->AllocatorFree(allocator, inputName));
+    if (code != ORT_OK) {
+      // break out on error, return array and let Java throw the exception.
```

Comment:
>  // break out [](http://example.com/codeflow?start=4&length=14)

Is Java Array going to be GCed?

Reply:
Yes, it's a real Java object but has no references so will be cleaned up the next time the GC runs after the JNI function returns.